--- scripts/bootstrap/compile.sh.orig	2023-10-22 21:05:54.513270279 +0200
+++ scripts/bootstrap/compile.sh	2023-10-22 20:59:51.287464036 +0200
@@ -445,6 +445,7 @@
   "${JAVA_HOME}/bin/java" \
       -XX:+HeapDumpOnOutOfMemoryError -Xverify:none -Dfile.encoding=ISO-8859-1 \
       -XX:HeapDumpPath=${OUTPUT_DIR} \
+      --add-opens java.base/java.lang=ALL-UNNAMED \
       -Djava.util.logging.config.file=${OUTPUT_DIR}/javalog.properties \
       ${JNI_FLAGS} \
       -jar ${ARCHIVE_DIR}/libblaze.jar \
